Encoder/Decoder new Data Types support

ronfarkash opened the pull request:
- [x] Unsigned VarInt 
- [x] Double
- [x] UVarIntBytes (Compact Bytes)
- [x] UVarIntString (Compact String)
- [x] UVarIntArray (Compact Array)

Double is missing unit tests, the rest are missing implementation.

A solution to this #1048, still incomplete since I am unsure about these things:

- should I stack the encoder.spec.js with more tests for each data type I implement?
- in decoder.js the function `readUVarInt()` which was taken from the same place ([stream-lib/Varint.java](https://github.com/addthis/stream-lib/blob/master/src/main/java/com/clearspring/analytics/util/Varint.java#L183)) the function `writeUVarInt()` was taken from.

    The function throws an error if you try to decode a UVarInt with more than 5 bytes, is that fine to keep it throw statement there?

    It also returns the value after unsigned right shift operation due to JavaScript's behavior with numbers. I explained it above the function but I am not sure if the [explanation](https://stackoverflow.com/a/1822769/6314048) should stay there?
    
    stream-lib does **not**  use [VLQ](https://en.wikipedia.org/wiki/Variable-length_quantity) they use [LEB128](https://en.wikipedia.org/wiki/LEB128) format, I am mentioning this since when I wrote the unit tests for the functions I instantly assumed VLQ was used and hit my head against the wall for a day wondering why the examples from the wiki give me different results so I added a comment above the function `writeVarInt()` stating that LEB128 is used. Should the comment stay there?

- Keeping naming convention encoder.js/decoder.js blind to Java. 
  Java & Kafka calls the last 3 data types in the list "Compact ???" due to the nature of LEB128 shrinking the request size in some situations, should the function names be for example `writeCompactString()` or `writeUVarString()`?

- Some files like `jsconfig.json` and `yarn.lock` were added to this branch I am not sure what are the consequences of that but I assume that shouldn't happen.

Implementation isn't done and I am sure some of the things I listed are trivial but I would like to get a review for this 😊.

Edit 1:
I see that I did start implementing `writeUVarIntArray()`, naming question still remains.

Nevon commented on src/protocol/decoder.js:
```diff
@@ -196,6 +203,23 @@ module.exports = class Decoder {
+      }
+    }
+    result |= currentByte << i
+    return result >>> 0
```
While I do understand that the unsigned right shift is here to make sure that you round the number to a value between 0 and 2 ** 32 - 2, I don't really understand why it's necessary. Won't all valid unsigned varints fit in that range kind of by definition? And if somehow we get a number that's not in that range, is rounding the right thing to do? I really don't know.

ronfarkash replied:
While running the boundaries test, the top boundary test would receive a `-1` instead of `4294967295`
`expect(decode32u(unsigned32(MAX_SAFE_UNSIGNED_INT))).toEqual(MAX_SAFE_UNSIGNED_INT)`

I was confused so I double checked it in Python and saw something was off. I looked it up online and saw a very detailed explaination to why that happens, to sum it up (straight from [here](https://stackoverflow.com/a/1822769)):

JS Numbers are double-precision floats, bitwise operators convert these numbers to 32-bit signed integers so anything above `2^31` would come back around as negative. To fix that I need a way to convert my result to fit unsigned integers. 

Luckily `>>> 0` converts the results of it's calculation to an unsigned integer and ensures my integer is between the range `0` and `0xFFFFFFFF`

Nevon reviewed (changes requested):
Very nice start!

Nevon commented:
> should I stack the encoder.spec.js with more tests for each data type I implement?

Yes please. If it becomes unmanageably large we can start to split it apart.

> The function throws an error if you try to decode a UVarInt with more than 5 bytes, is that fine to keep it throw statement there?

Yup. Just change it to throw a custom error instead of a generic `Error`.

> It also returns the value after unsigned right shift operation due to JavaScript's behavior with numbers. I explained it above the function but I am not sure if the explanation should stay there?

I would prefer it if we can write a test that shows why it's necessary (i.e. it fails if we don't round the value to an int32) and has a descriptive name, but otherwise a comment is fine. The kind of code in Encoder is the kind where it's just about impossible to read and understand why certain things are done, so the more context we can preserve, the better.

> Keeping naming convention encoder.js/decoder.js blind to Java.
Java & Kafka calls the last 3 data types in the list "Compact ???" due to the nature of LEB128 shrinking the request size in some situations, should the function names be for example writeCompactString() or writeUVarString()?

Good question! Given that we already have for example `writeVarIntString` I would stay consistent with that and make it `writeUVarIntString`.

ronfarkash commented on src/protocol/decoder.js:
```diff
@@ -196,6 +203,23 @@ module.exports = class Decoder {
+    while (((currentByte = this.buffer[this.offset++]) & MOST_SIGNIFICANT_BIT) !== 0) {
+      result |= (currentByte & OTHER_BITS) << i
+      i += 7
+      if (i > 35) {
```
You're right, I mindlessly took it from stream-lib without noticing this.

ronfarkash replied:
Should be fixed next push.

ronfarkash commented on jsconfig.json:
```diff
@@ -0,0 +1,14 @@
+{
```
Removed the file.

ronfarkash commented on src/protocol/decoder.js:
```diff
@@ -196,6 +203,23 @@ module.exports = class Decoder {
+      result |= (currentByte & OTHER_BITS) << i
+      i += 7
+      if (i > 35) {
+        throw new Error('Variable length quantity is too long')
```
Done.

ronfarkash commented:
> I would prefer it if we can write a test that shows why it's necessary (i.e. it fails if we don't round the value to an int32) and has a descriptive name, but otherwise a comment is fine. The kind of code in Encoder is the kind where it's just about impossible to read and understand why certain things are done, so the more context we can preserve, the better.

I couldn't think of a test to showcase it's necessity. I believe a comment will save up some time for people who want to get to know the Encoder, the comment above `writeVarInt()` (that I edited a bit) stating where the code was originally taken from helped me a ton.

Regarding the new function I added `writeUVarIntString()` representing compact strings, as stated in the [protocol guide](https://kafka.apache.org/protocol#protocol_types) "Represents a sequence of characters. **First the length N + 1 is given as an UNSIGNED_VARINT** . Then N bytes follow which are the UTF-8 encoding of the character sequence."

I don't know why Kafka decided to encode the length + 1, but I implemented the function with it:
```js
this.writeUVarInt(byteLength + 1)
```

Nevon commented:
> I couldn't think of a test to showcase it's necessity. I believe a comment will save up some time for people who want to get to know the Encoder,

Totally fine, and I agree. I was mostly talking about the general case. If we can explain something with a test, that's ideal, but when that's not doable or when it doesn't make it clear enough, a comment is great.

Nevon reviewed (changes requested):
The implementation looks great - the only thing is that the yarn lockfile has changed, for some reason, which it should not have unless the package.json has a corresponding change to explain it. Remove that change from your PR, and we should be good to go.

Nevon commented:
Thanks for the contribution!
